pkg/vpcmodel/synthesisOutput.go

```diff
@@ -0,0 +1,153 @@
+)
+
+type SynthesisOutputFormatter struct {
+}
+
+func (j *SynthesisOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
+	conn *VPCConnectivity,
+	subnetsConn *VPCsubnetConnectivity,
+	cfgsDiff *diffBetweenCfgs,
+	outFile string,
+	grouping bool,
+	uc OutputUseCase,
+	explanation *Explanation, detailExplain bool) (*SingleAnalysisOutput, error) {
+	var all interface{}
+	switch uc {
+	case AllEndpoints:
+		all = GetSpec(conn.GroupedConnectivity.GroupedLines)
+	case AllSubnets:
+		all = GetSpec(subnetsConn.GroupedConnectivity.GroupedLines)
+	}
+	outStr, err := writeJSON(all, outFile)
+	v2Name := ""
+	if c2 != nil {
+		v2Name = c2.VPC.Name()
+	}
+	return &SingleAnalysisOutput{Output: outStr, VPC1Name: c1.VPC.Name(), VPC2Name: v2Name, format: Synthesis, jsonStruct: all}, err
+}
+
+func handleExternals(srcName, cidrOrAddress string, externalsMap map[string]string, externals spec.SpecExternals) string {
+	if val, ok := externalsMap[srcName]; ok {
+		return val
+	}
+	name := "external" + strconv.Itoa(len(externals))
+	externalsMap[srcName] = name
+	externals[name] = cidrOrAddress
+	return name
+}
+
+func handleNameAndType(resource EndpointElem, externalsMap map[string]string, externals spec.SpecExternals) (
+	resourceName string,
+	resourceType spec.ResourceType) {
+	resourceName, nifNumber := resource.DetailedResourceForSynthesisOut() // for synthesis output return two
+	if resource.IsExternal() {
+		if structObj, ok := resource.(*groupedExternalNodes); ok {
+			// should be always true if src is external
+			resourceName = handleExternals(resourceName, structObj.CidrOrAddress(), externalsMap, externals)
+		}
+	}
+	resourceType = resource.SynthesisKind()
+
+	// if this nif's vsi has only one nif, we convert it to instance type with name of the instance
+	// because the name of the nif will be meaningless for the user if there is one generated nif.
+	if resourceType == spec.ResourceTypeNif && nifNumber == 1 {
+		resourceType = spec.ResourceTypeInstance
+	}
+	return
+}
+
+func GetSpec(groupedLines []*groupedConnLine) spec.Spec {
+	s := spec.Spec{}
+	connLines := []spec.SpecRequiredConnectionsElem{}
+	externals := spec.SpecExternals{}
+	externalsMap := make(map[string]string)
+	sortGroupedLines(groupedLines)
+
+	for _, groupedLine := range groupedLines {
+		srcName, srcType := handleNameAndType(groupedLine.Src, externalsMap, externals)
+		dstName, dstType := handleNameAndType(groupedLine.Dst, externalsMap, externals)
+		if groupedLine.CommonProperties.Conn.isEmpty() {
+			continue
+		}
+		connLines = append(connLines, spec.SpecRequiredConnectionsElem{
+			Src:              spec.Resource{Name: srcName, Type: srcType},
+			Dst:              spec.Resource{Name: dstName, Type: dstType},
+			AllowedProtocols: sortProtocolList(spec.ProtocolList(connection.ToJSON(groupedLine.CommonProperties.Conn.allConn)))})
+	}
+	s.Externals = externals
+	s.RequiredConnections = connLines
+	return s
+}
+
+func sortProtocolList(g spec.ProtocolList) spec.ProtocolList {
```

Comment:
Any reason to do it so precisely?
How about a simpler sorting scheme: marshal each protocol into string, and sort the strings.
```golang
	sort.Slice(g, func(i, j int) bool {
		iMarshal, _ := json.Marshal(g[i])
		jMarshal, _ := json.Marshal(g[j])
		return string(iMarshal) > string(jMarshal)
	})

```

Reply:
no reason, will change it